add profile links from other communities

cellio opened the pull request:
Minimally addresses https://github.com/codidact/qpixel/issues/284 by adding an "all communities" tab to the user profile to hold links to that user's profile on all other network communities where a profile exists. 

Here's a moderator viewing its own tab:

![tab with table: community profile link, mod status where applicable, number of posts, rep](https://github.com/user-attachments/assets/0bc8f583-b270-4c14-96a5-18a932510b58)

And here's how someone else's profile looks:

![same table columns, fewer tabs](https://github.com/user-attachments/assets/07b10e95-7c3c-48d7-bfa8-a785c050192e)

The guts of the profile tab are separated from the "container" (`network.html.erb` renders `_network.html.erb`) to make it easier to pick this up and move it to a truly global place later if we can work out how to do that.  (See discussion on the linked issue.)  Right now, your network profile is part of your user page on each specific community, just like your inbox is, even though these things are really "higher up".

To support this, I added several "accessor" functions to the user model to go fetch information on specific communities.  I followed the pattern of the pre-existing `has_ability_on` function.

Currently this table shows the number of (undeleted) posts on each community.  I was going to separate that out by post type, but post types can be defined in configuration, so that means walking the list of all post types to get their names and counts on each community (some of which might not be enabled on some communities), and I decided to defer that until we decide we need it.  If the consensus from user feedback is that we should have that, we can do it as followup work.

This initial implementation does nothing special with sorting, and I believe the order will automatically be the same as the site switcher and the dashboard.  This implementation also does not give the user a way to change the sort order.  A full implementation would be user-customizable, which requires storing settings somewhere.  Also, we need to work out the UI for custom ordering.

This implementation does not yet have any user-private information.  I once proposed a design that would show you things like where you have new inbox items or flag responses.  I'd still like to do that, but even without it, I think a basic list is better than what we have today.  As a moderator I have sometimes needed to look at a user's activity on other communities, and neither URL editing nor navigating the users list is satisfactory.

I learned a lot about Ruby/Rails while doing this and probably failed to learn other things.  I suspect there are stylistic anomalies in my code ("no don't do it _that_ way!"), which I hope reviewers will point out so I can fix them.

ws909 commented:
For bookkeeping, I’d suggest renaming this to «Listing profile information from across the network», since it isn’t actually a network-profile.

Oaphi commented on app/helpers/users_helper.rb:
```diff
@@ -61,10 +61,13 @@ def rtl_safe_username(user)
+      link_to 'deleted user', '#', { dir: 'ltr' }.merge(link_opts).except(:anchortext)
     else
-      link_to user.rtl_safe_username, user_url(user, **url_opts), { dir: 'ltr' }.merge(link_opts)
+      link_to user.rtl_safe_username, user_url(user, **url_opts), { dir: 'ltr' }.merge(link_opts).except(:anchortext)
```
Optional: let's DRY this a little and extract `{ dir: 'ltr' }.merge(link_opts).except(:anchortext)` into a variable available for all branches and pass it instead.

Oaphi replied:
Optional: let's DRY this a little and extract `{ dir: 'ltr' }.merge(link_opts).except(:anchortext)` into a variable available for all branches and pass it instead.

Oaphi commented on app/models/user.rb:
```diff
@@ -138,6 +138,31 @@ def is_admin
+
+  def post_count_on(community_id)
+    cu = community_users.where(community_id: community_id).first
+    cu&.post_count
```
Optional: let's be extra sure and return `0`  by default to avoid accidental bugs with consumers of the API doing a safe navigation check. Has an additional benefit of preventing "type hopping" by providing a stable API contract to consumers ("always returns an integer", etc).

Oaphi commented on app/models/user.rb:
```diff
@@ -138,6 +138,31 @@ def is_admin
+
+  def reputation_on(community_id)
+    cu = community_users.where(community_id: community_id).first
+    cu&.reputation
```
Optional: same as with the `post_count_on` and `reputation_on` suggestions, let's default to `1` here in case `cu` fails the safe navigation check.

Oaphi commented on app/models/user.rb:
```diff
@@ -138,6 +138,31 @@ def is_admin
+  # Used by network profile: does this user have a profile on that other comm?
+  def has_profile_on(community_id)
+    cu = community_users.where(community_id: community_id).first
+    !cu&.user_id.nil?
```
Optional: same as with the `post_count_on` and `reputation_on` suggestions, let's default to `false` if `cu` fails the safe navigation check.

cellio replied:
Heh.  For this line I was rubocopped -- I was more explicit originally. :-)

Oaphi commented on app/models/user.rb:
```diff
@@ -138,6 +138,31 @@ def is_admin
+
+  def is_moderator_on(community_id)
+    cu = community_users.where(community_id: community_id).first
+    if cu&.is_moderator || cu&.privilege?('mod')
+      true
+    else
+      false
+    end
```
`if ... true else false` is redundant and should never be done like that, you can simply return the boolean value directly: `cu&.is_moderator || cu&.privilege?('mod')` (that's it, without the whole construct).

Oaphi reviewed (changes requested):
Other than a single redo item (the rest are optional suggestions), LGTM at a quick glance!

cellio commented:
> For bookkeeping, I’d suggest renaming this to «Listing profile information from across the network», since it isn’t actually a network-profile.

When you say "renaming", where do you mean?

ws909 commented:
> When you say "renaming", where do you mean?

PR title; currently named "basic network profile".

cellio commented:
> Other than a single redo item (the rest are optional suggestions), LGTM at a quick glance (at least from the initial implementation standpoint)!

I addressed your comments -- with luck, in the ways you intended.

Oaphi reviewed (approved):
Leaving for others to take a quick look too, but otherwise - LGTM

ArtOfCode- commented on app/models/community_user.rb:
```diff
@@ -24,6 +24,11 @@ def suspended?
 
+  # All undeleted posts on this community by this user.
+  def post_count
+    Post.unscoped.where(community_id: community_id).where(user: user).undeleted.count
```
This is going to be expensive when run for multiple users and/or communities on one page. We're _probably_ alright at current scale, but we should ultimately cache this on the CommunityUser model for each community.

cellio replied:
It occurs to me that the community-specific post count shows up other places, including on the profile, where we call `@user.posts.undeleted.count`.  But I'm not quite sure how to do that for a non-local user -- any advice?  We're calling this from a user on site A and asking about the user on site B, C, D... .

cellio replied:
Filed https://github.com/codidact/qpixel/issues/1411 after discussion with @ArtOfCode- .
